allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+                instance = self.text_to_instance(
+                    question,
+                    tokenized_question,
+                    context,
+                    tokenized_context_window,
+                    answers,
+                    window_token_answer_span,
+                    additional_metadata,
+                )
+                yield instance
+
+            stride_start += space_for_context
+            if stride_start >= len(tokenized_context):
+                break
+            stride_start -= self.stride
+
+    @overrides
+    def text_to_instance(
+        self,  # type: ignore
+        question: str,
+        tokenized_question: List[Token],
+        context: str,
+        tokenized_context: List[Token],
+        answers: List[str],
+        token_answer_span: Optional[Tuple[int, int]],
+        additional_metadata: Dict[str, Any] = None,
+    ) -> Instance:
+        fields = {}
+
+        # make the question field
+        def cls_token():
+            return Token(
+                self._tokenizer.tokenizer.cls_token,
+                text_id=self._tokenizer.tokenizer.cls_token_id,
+                type_id=self.non_content_type_id,
+            )
+
+        def sep_token():
+            return Token(
+                self._tokenizer.tokenizer.sep_token,
+                text_id=self._tokenizer.tokenizer.sep_token_id,
+                type_id=self.non_content_type_id,
+            )
+
+        question_field = TextField(
+            (
+                [cls_token()]
+                + tokenized_question
+                + [sep_token(), sep_token()]
+                + tokenized_context
+                + [sep_token()]
+            ),
+            self._token_indexers,
+        )
+        fields["question"] = question_field
+        start_of_context = 1 + len(tokenized_question) + 2
+
+        # make the answer span
+        if token_answer_span is not None:
+            assert all(i >= 0 for i in token_answer_span)
+            assert token_answer_span[0] < token_answer_span[1]
+
+            fields["span_start"] = IndexField(
+                token_answer_span[0] + start_of_context, question_field
+            )
+            fields["span_end"] = IndexField(token_answer_span[1] + start_of_context, question_field)
+
+            assert all(
+                t.type_id == 0
+                for t in question_field.tokens[
+                    fields["span_start"].sequence_index : fields["span_end"].sequence_index
+                ]
+            )
+        else:
+            fields["span_start"] = IndexField(-1, question_field)
+            fields["span_end"] = IndexField(-1, question_field)
+
+        # make the context span, i.e., the span of text from which possible answers should be drawn
+        fields["context_start"] = IndexField(start_of_context, question_field)
+        fields["context_end"] = IndexField(
+            start_of_context + len(tokenized_context), question_field
+        )
+        assert (
```

Comment:
This one is definitely not checking the input data, as it's something that you produced yourself.  Right?  If so, I'd definitely recommend this be a test instead of an assert.  Or are you worried about what someone might pass in in a demo?

Reply:
Asserts are not substitutes for tests. They are documentation that never goes stale.

Reply:
Tests are also documentation that never goes stale, and that documentation doesn't get in the way of reading through the logic of a method.  But, this isn't something I want to argue with you about; just do what you think is best.  One suggestion, though - if you think the asserts are useful, having meaningful error messages attached to them would be nice.

Reply:
Yeah, one of my gripes with Python (or, rather, how it's commonly written) is that it conflates code invariant checks, a.k.a. asserts, with input validation. If a clear distinction is made between those two cases you can have some pretty heavyweight asserts, but turn them off in performance critical situations. We can't do that because we use `assert` for input validation as well. And even if we changed our code it's likely other packages have that issue.

Given all of that, I'd suggest we allow `assert` for lightweight code invariant checks. If the assertion is expensive, however, we'd have to rely on a test as Matt says.

Reply:
My goal is not to "disallow" anything; I think the use of asserts is a stylistic decision that I don't want to try to impose on people.  But my motivation here has nothing to do with how heavy an assertion is, it's about readability and what object has the responsibility to enforce contracts.

Asserts in general hurt readability, as they distract from the logic that's being performed.  Occasionally they're worth it anyway.

Asserts that do input (or output) validation, when we own the logic that's producing the input (or output), seem like they are in the wrong place.  They say that we're delegating checking behavior from one object (that we own!) to all of its users.  That seems suboptimal - instead of one set of tests, we now have asserts everywhere a function is called.  There's definitely a place for input validation on methods that we expect code we don't own to call, but in python those are typically `ValueErrors`, not asserts.

Reply:
Sorry, @matt-gardner. I think my comment may have come across poorly. I wasn't trying to put words in your mouth, just generally make some comments about my view of assertions that could be relevant.

I suspect there is some philosophical disagreement here regarding assertions, but let's chat about that over lunch sometime. :)

Reply:
Sounds like you read my comment as stronger than I intended it - nothing came across poorly to me.  Yes, there is some philosophical disagreement about asserts, which is why I'm trying not to be picky.  There are a few things I think we should definitely standardize in the library, though:

1. When doing input validation, we should use `ValueError` instead of `assert` - that's what `ValueError` is for in python.
2. When we are calling functions that we have implemented, we should use tests on that function instead of asserts in the calling code.  See my comments in the model in this PR, where there are asserts checking the output of `get_best_span`.

For all other uses of asserts, I will not be picky.

Reply:
Thanks for clarifying. I strongly agree with 1) and generally agree with 2). There may be a small number of cases where a postcondition assert in the _called_ function is appropriate in addition to a test, but I don't mean to advocate for that in the general case and would be happy for it to be subject to extra scrutiny during review.

Reply:
I moved the `type_id` checking into a test: 35525021ebdd7b459186b0abe22b8122cdca223e

Reply:
We are checking input data with asserts? That's out of scope for this PR, but that doesn't seem like a good thing. That's what `ValueError` is for.

Reply:
I don't know whether that's happening anywhere (checking input data with asserts); I mentioned it just because @brendan-ai2 had it in a prior comment.  I think we're all in agreement on this point.

Reply:
Woha. When I typed up my response, half of your discussion wasn't on my screen. I did not respond to what I thought I responded to. Not to self, refresh the page before typing.